File: app/src/main/java/com/zzp/dtrip/activity/RegisterActivity.java
package com.zzp.dtrip.activity;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.RadioButton;
import android.widget.TextView;
import android.widget.Toast;

import com.zzp.dtrip.R;
import com.zzp.dtrip.util.HttpBeanService;
import com.zzp.dtrip.javabean.RegisterAsk;
import com.zzp.dtrip.javabean.RegisterResponse;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RegisterActivity extends AppCompatActivity implements View.OnClickListener {
    private TextView accountTextview;
    private TextView passwordTextview;
    private TextView password2Textview;
    private Button registerButton;
    private RadioButton manButton;
    private RadioButton womenButton;
    private Retrofit retrofit;
    private HttpBeanService httpBeanService;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);
        accountTextview =findViewById(R.id.account_register);
        passwordTextview =findViewById(R.id.password_register);
        password2Textview =findViewById(R.id.password2_register);
        registerButton=findViewById(R.id.register_button);
        registerButton.setOnClickListener(this::onClick);
        manButton=findViewById(R.id.register_man);
        manButton.setOnClickListener(this::onClick);
        womenButton=findViewById(R.id.register_women);
        womenButton.setOnClickListener(this::onClick);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId())
        {
            case R.id.register_button:register();break;
            case R.id.register_man:manButton.setChecked(true);womenButton.setChecked(false);break;
            case  R.id.register_women:manButton.setChecked(false);womenButton.setChecked(true);break;
        }

    }

    private void register() {
        String account = accountTextview.getText().toString();
        String password = null;
        String sex = "0";
        if (password2Textview.getText().toString() == passwordTextview.getText().toString())
            password = passwordTextview.getText().toString();
        else {
            Toast.makeText(RegisterActivity.this, "密码不一致，请重试", Toast.LENGTH_SHORT).show();

        }

        if (manButton.isChecked() == true)
            sex = "0";
        if (womenButton.isChecked() == true)
            sex = "1";
        RegisterAsk ask = new RegisterAsk();
        ask.setAccount(account);
        ask.setPassword(password);
        ask.setSex(sex);
        retrofit = new Retrofit.Builder().baseUrl("http://101.34.85.209.5240/user/")
                .addConverterFactory(GsonConverterFactory.create()) // 设置数据解析器
                .build();
        httpBeanService = retrofit.create(HttpBeanService.class);
        Call<RegisterResponse> call = httpBeanService.register(ask);
        call.enqueue(new Callback<RegisterResponse>() {
            @Override
            public void onResponse(Call<RegisterResponse> call, Response<RegisterResponse> response) {
            }

            @Override
            public void onFailure(Call<RegisterResponse> call, Throwable t) {

            }
        });
    }

    }

File: app/src/main/java/com/zzp/dtrip/activity/RegisterActivity.kt
package com.zzp.dtrip.activity

import android.os.Bundle
import android.util.Log
import android.widget.RadioGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.button.MaterialButton
import com.google.android.material.radiobutton.MaterialRadioButton
import com.google.android.material.textfield.TextInputLayout
import com.zzp.dtrip.R
import com.zzp.dtrip.body.RegisterBody
import com.zzp.dtrip.data.NormalResult
import com.zzp.dtrip.util.AppService
import com.zzp.dtrip.util.RetrofitManager
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class RegisterActivity : AppCompatActivity() {

    private lateinit var usernameLayout: TextInputLayout
    private lateinit var passwordLayout: TextInputLayout
    private lateinit var passwordAgainLayout: TextInputLayout
    private lateinit var registerButton: MaterialButton
    private lateinit var manButton: MaterialRadioButton
    private lateinit var womanButton: MaterialRadioButton
    private lateinit var radioGroup: RadioGroup

    private var username = ""
    private var password = ""
    private var passwordAgain = ""
    private var sex = ""

    private val TAG = "RegisterActivity"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)
        findViewById()
        setFocus()

        registerButton.setOnClickListener {
            username = usernameLayout.editText?.text.toString()
            password = passwordLayout.editText?.text.toString()
            passwordAgain = passwordAgainLayout.editText?.text.toString()
            if (isNotEmpty()) {
                if (password != passwordAgain) {
                    passwordAgainLayout.error = "密码不一致"
                }
                else {
                    postRegister()
                }
            }
        }

        radioGroup.setOnCheckedChangeListener { group, checkedId ->
            when(checkedId) {
                R.id.register_man -> {
                    manButton.setTextColor(resources.getColor(R.color.white))
                    womanButton.setTextColor(resources.getColor(R.color.gray))
                    sex = "0"
                }
                R.id.register_woman -> {
                    womanButton.setTextColor(resources.getColor(R.color.white))
                    manButton.setTextColor(resources.getColor(R.color.gray))
                    sex = "1"
                }
            }
        }
    }

    private fun findViewById() {
        usernameLayout = findViewById(R.id.username_layout)
        passwordLayout = findViewById(R.id.password_layout)
        passwordAgainLayout = findViewById(R.id.password2_layout)
        registerButton = findViewById(R.id.register_button)
        manButton = findViewById(R.id.register_man)
        womanButton = findViewById(R.id.register_woman)
        radioGroup = findViewById(R.id.radio_group)
    }

    private fun postRegister() {
        val appService = RetrofitManager.create<AppService>()
        val sexNum = if (sex == "男") 0 else 1
        val task = appService.postRegister(RegisterBody(username, password, sexNum.toString()))
        task.enqueue(object : Callback<NormalResult>{
            override fun onResponse(call: Call<NormalResult>,
                                    response: Response<NormalResult>) {
                Log.d(TAG, "onResponse: ")
                response.body()?.apply {
                    if (errorCode == 0) {
                        Toast.makeText(this@RegisterActivity,
                            "注册成功", Toast.LENGTH_SHORT).show()
                        onBackPressed()
                    }
                    else {
                        Toast.makeText(this@RegisterActivity,
                            errorMsg, Toast.LENGTH_SHORT).show()
                    }
                }
            }

            override fun onFailure(call: Call<NormalResult>, t: Throwable) {
                Log.d(TAG, "onFailure: $t")
            }
        })
    }


    private fun setFocus() {
        usernameLayout.editText?.setOnFocusChangeListener { v, hasFocus ->
            usernameLayout.error = ""
        }
        passwordLayout.editText?.setOnFocusChangeListener { v, hasFocus ->
            passwordLayout.error = ""
        }
        passwordAgainLayout.editText?.setOnFocusChangeListener { v, hasFocus ->
            passwordAgainLayout.error = ""
        }
    }

    private fun isNotEmpty() : Boolean {
        var flag = true
        if (username.trim().isEmpty()) {
            usernameLayout.error = "用户名为空"
            flag = false
        }
        else {
            usernameLayout.error = ""
        }
        if (password.trim().isEmpty()) {
            passwordLayout.error = "密码为空"
            flag = false
        }
        else {
            passwordLayout.error = ""
        }
        if (passwordAgain.trim().isEmpty()) {
            passwordAgainLayout.error = "密码为空"
            flag = false
        }
        else {
            passwordAgainLayout.error = ""
        }
        return flag
    }
}